product-scenarios/code-coverage/code-coverage.sh

```diff
@@ -0,0 +1,164 @@
+        ipaddress=$(echo `curl http://169.254.169.254/latest/meta-data/public-ipv4`)
+
+        #Get Active Agents for the defined test plan
+        mkdir -p "${HOME}"/code-coverage/resources
+        echo $(curl -X GET "${VAR_TINKERER_ENDPOINT}"/api/test-plan/"${VAR_TEST_PLAN_ID}"/agents \
+        -u "${VAR_TINKERER_USERNAME}":"${VAR_TINKERER_PASSWORD}") > "${HOME}"/code-coverage/resources/agentsList.json
+
+        python -m json.tool "${HOME}"/code-coverage/resources/agentsList.json | \
+        awk -F'"' '/instanceName/{print $4}' > "${HOME}"/code-coverage/resources/agent_name_list.txt
+
+        i=0
+        while read line
+        do
+            agent_name[$i]="$line"
+
+            #Stop the wso2server.sh
+            echo "Stop the server in node ${agent_name[i]}"
+            
+            product=$(basename "${VAR_REM_DIR}" | cut -d- -f1)
+            
+            if [ "$product" = "wso2esb" ]; then
+                echo $(tinkerer_curl_commands ${agent_name[i]} "$VAR_REM_DIR/bin/wso2server.sh\tstop")
+            else
+                echo $(tinkerer_curl_commands ${agent_name[i]} "$VAR_REM_DIR/bin/integrator.sh\tstop")
+            fi
+            sleep 30
+
+            #Zip the jacoco folder
+            echo "Zip the jacoco folder in node ${agent_name[i]}"
+
+            echo $(tinkerer_curl_commands ${agent_name[i]} "cd\t$VAR_REM_DIR/repository/logs/jacoco;\tzip\t-r\tjacoco.zip\t.")
+
+            #Generate key
+            echo "Generate a key in node ${agent_name[i]}"
+
+            echo $(tinkerer_curl_commands ${agent_name[i]} "mkdir\t-p\t$VAR_REM_DIR/../keys")
+            echo $(tinkerer_curl_commands ${agent_name[i]} "ssh-keygen\t-b\t2048\t-t\trsa\t-f\t$VAR_REM_DIR/../keys/deploy.key\t-q\t-N\t''")
+
+            #Add key to authorized_keys
+            echo "Add key of node ${agent_name[i]} to the authorized_keys in client side"
+
+            echo `tinkerer_curl_commands ${agent_name[i]} "cat\t$VAR_REM_DIR/../keys/deploy.key.pub"` | \
+            sed -nE 's/.*"response":"(.*)","completed.*/\1/p' >> /home/ubuntu/.ssh/authorized_keys
+
+            #Copy file to client side
+            echo "Copy file to the client side from ${agent_name[i]}"
+            mkdir -p ${HOME}/code-coverage/resources/instance$((i+1))
+
+            #Give permission to .key file
+            tinkerer_curl_commands ${agent_name[i]} "sudo\tchmod\t400\t$VAR_REM_DIR/../keys/deploy.key"
+
+            echo $(tinkerer_curl_commands ${agent_name[i]} \
+            "scp\t-o\tStrictHostKeyChecking=no\t-i\t$VAR_REM_DIR/../keys/deploy.key\t$VAR_REM_DIR/repository/logs/jacoco/jacoco.zip\t$user@$ipaddress:$HOME/code-coverage/resources/instance$((i+1))")
+            sleep 30
+
+            echo "Extract the copied coverage artifact file of ${agent_name[i]}"
+            unzip ${HOME}/code-coverage/resources/instance$((i+1))/jacoco.zip \
+            -d ${HOME}/code-coverage/resources/instance$((i+1))/jacoco
+
+            sleep 40
```

Comment:
Currently we cannot use waits to wait until command execution complete, because using tinkerer we open up a new shell and execute the provided command in server side. Therefore used multiple sleeps to wait.